c/cert/src/rules/ENV34-C/DoNotStorePointersReturnedByCertainFunctions.ql

```diff
@@ -0,0 +1,43 @@
+/**
+ * @id c/cert/do-not-store-pointers-returned-by-certain-functions
+ * @name ENV34-C: Do not store pointers returned by certain functions
+ * @description The value returned by functions `getenv`, `asctime`, `localeconv`, `setlocale`, and
+ *              `strerror` may be overwritten by a second call.
+ * @kind problem
+ * @precision very-high
+ * @problem.severity error
+ * @tags external/cert/id/env34-c
+ *       correctness
+ *       external/cert/obligation/rule
+ */
+
+import cpp
+import codingstandards.c.cert
+import semmle.code.cpp.dataflow.DataFlow
+
+/*
+ * A model of environment functions
+ */
+
+class GetEnvFunctionCall extends FunctionCall {
+  GetEnvFunctionCall() {
+    this.getTarget().getName() = ["getenv", "asctime", "localeconv", "setlocale", "strerror"]
+  }
+}
+
+from GetEnvFunctionCall fc1, GetEnvFunctionCall fc2, Expr e
+where
+  not isExcluded(e, Contracts2Package::doNotStorePointersReturnedByCertainFunctionsQuery()) and
+  // A variable `v` is assigned with `GetEnvFunctionCall` return value
+  exists(Variable v |
+    fc1 = v.getInitializer().getExpr()
+    or
+    exists(Assignment ae | fc1 = ae.getRValue() and v = ae.getLValue().(VariableAccess).getTarget())
+  ) and
+  // A second `GetEnvFunctionCall`
+  fc2 = fc1.getASuccessor+() and
+  // The value is is accessed in `e` afer the second `GetEnvFunctionCall`
+  e = fc2.getASuccessor+() and
```

Comment:
This approach doesn't take into account definition reachability, which means it will be confused by variable reuse:
```cpp
  char* temp = getEnv("VAR1");
  use(temp);
  temp = getEnv("VAR2");
  use(temp);
```
And also loops:
```cpp
for(std:string envVar : envVars) {
  char* temp = getEnv(envVar);
  use(temp);
}
```
The usual fix for this type of reachability problem is to use `StackVariableReachability`, however, you want to assert that the use is via a second call to the same function, which is not supported.

Furthermore, I think the overall approach risks missing results where the repetition is non-local - i.e. stored in a field or global variable.

Therefore, my suggestion would be to flag _all_ cases where the result is stored in a variable, and then use a `StackVariableReachability` to exclude simple cases where no duplicate calls can occur. You need to be a bit careful here - any call to a function could _indirectly_ call the same function, and I think we should err on the side of reporting results in that case.




If you want to implement this approach, I suggest using `StackVariableReachabilityWithReassignment` to track

Reply:
Ah, apologies - using dataflow does indeed address the variable reuse case, but not the loop case.

> I added a second version of the query using StackVariableReachabilityWithReassignment to remove false positives due to variable rewrite.

I actually intended something more drastic here - that we flag every assignment of a returned pointer as an alert, and only consider it "safe" if we are convinced that every use of the assigned variable is safe.

The `StackVariableReachabilityWithReassignment` could then be used to determine what uses are "safe", i.e. find def-use pairs which don't go through a direct or indirect call to the problematic APIs. If all uses of the assignment are safe, and the pointer doesn't escape the function after the function ends, then we don't flag it, otherwise we do.

`StackVariableReachabilityWithReassignment` doesn't work with global variables or fields, so we  would have to take a different approach for those. I do think it's pretty dubious to store any returned pointers in global variables or fields, though - it just seems to be asking for trouble.

I don't if overall this approach would have too many false positives. However, I do think the current approach could have too many false negatives (with globals, fields etc.) and we should try to capture those as well.